### DISCUSSION TITLE
Panel creation enhancement idea

### DISCUSSION THREAD

**Discussion by Psi58** [GrandOrgue/GoOdf#76] [Ideas]
GOODF is very useful - the GUI gives both a high-level understanding of an organ and enables low-level tinkering in the context of the whole instrument. Thank you!

My main use for it, so far, has been to create left and right jambs for organs that don't provide them. Even with GOODF it's quite a laborious process, and could be made much slicker. In every case I have cropped the main console image to create the jambs. This means that the Left Jamb stop positioning coordinates are identical to the main console, but the Right Jamb coordinates need to be adjusted for a different origin.

Let's say I want to create a right jamb, here are the steps that I take (after creating the graphic and calculating the X-axis offset required):
1. Create the new panel, and create GUI elements from 'Available Organ Elements'
2. For each appropriate element in the Main Panel, right click and copy the attributes to the equivalent element in the new panel. Note the coordinates (PositionX and PositionY) of the element in the Main Panel and enter them, with the offset as required.
3. Repeat step 2 as many times as there are stops... For anything other than a very small organ this is tedious and can take quite a long time.

It seems to me that a lot of this time-consuming and repetitive work could be automated. The user should still calculate the X-offset. Then he/she should select all the required elements from the existing panel (ie Ctrl-click on them) in the <Panel Name> screen. Right-clicking should then open a dialog box allowing the offset to be entered, the new panel name selected - and then a 'Copy' button should create and populate everything else.

This should make panel creation really simple, and a trivial process. From my perspective as a naturally lazy person this would be very welcome!

**Comment by larspalo** [comment]
Thanks for testing GOODF! Your feature request is very specific - semi automatically create a right stop jamb. Personally I think that any panel copying feature should be a bit more flexible and have more use cases. For instance, a feature that allow to specify a rectangle on any panel (graphical rendition) and a further option to automatically create a new panel containing everything existing within that selected rectangle in exactly the same layout as if it was "cut out" from the original panel... It could include cutting the background images too as new images (if needed) so that it wouldn't be necessary to do manually.

However, this is not directly trivial to code, even if it's certainly possible. Expanding on your procedure description, if you perhaps could live with a little more sloppiness with regard to positioning, it would maybe be somewhat quicker to do like this:

1. Create a screenshot of the rendered panel (either in GOODF or in GO with all elements in off position) that you wish to split
2. Cut out the parts from the image you just created that and save them as new background(s) for the new panels in an image editor
3. Create the new panels in GOODF and add the newly created backgrounds as images
4. Add all GUI elements that should be on each new panel
5. Copy the attributes from the original panel GUI elements to the (same) GUI elements on the new panel (right click on a GUI element to open a dialog that allows that)
6. For each new panel open the graphical display of it in GOODF and drag each of the GUI elements to visually match the position they have on the background

Currently the copy feature of the GUI elements in GOODF doesn't include the positioning information - which is why you must add that part yourself either by dragging them to position or entering the position values - but it wouldn't be too much work to add options to that dialog that would include the positioning information either as is or with an offset.

**Reply by Psi58** [reply]
Many thanks, this would certainly make panel creation easier.

> "Your feature request is very specific - semi automatically create a right stop jamb. Personally I think that any panel copying feature should be a bit more flexible and have more use cases...."

Agreed, it's best to define the functionality so that it has wide appeal and value. I shared my specific (limited) experience, of course I recognise that others will use the tool in different ways. More "sloppiness" for a feature that's widely used is preferable to a 'perfect' solution for just one user :)

> "Add all GUI elements that should be on each new panel"

Just a note on this step - it's easy if the 'Available Organ Elements' are properly labelled and hence easy to identify. I was working on another organ a few months ago where the relevant field was empty, and I just had to guess... Maybe that was an unusual exception, and in most cases this approach will be fine.

> "For each new panel open the graphical display of it in GOODF and drag each of the GUI elements to visually match the position they have on the background..."

Optimising/automating this step would be a significant improvement. In my limited experience starting manually is unnecessary work - I have _always_ wanted to copy the layout from the main panel to the new panel, in which case just copying the coordinates (with user-definable offsets) would suffice. At the very least this would be a good starting point, after which further manual changes could still be made.
